examples_tests/01.HelloWorld/main.cpp

```diff
@@ -21,39 +28,282 @@ inline void debugCallback(nbl::video::E_DEBUG_MESSAGE_SEVERITY severity, nbl::vi
+			std::vector<VkExtensionProperties> availableExtensions(availableExtensionCount);
+			vkEnumerateDeviceExtensionProperties(gpu->getInternalObject(), NULL, &availableExtensionCount, availableExtensions.data());
+
+			bool requiredDeviceExtensionsAvailable = false;
+			for (uint32_t i = 0u; i < availableExtensionCount; ++i)
+			{
+				if (strcmp(availableExtensions[i].extensionName, requiredDeviceExtensionNames[0]) == 0)
+				{
+					requiredDeviceExtensionsAvailable = true;
+					break;
+				}
+			}
+
+			if (!requiredDeviceExtensionsAvailable)
+				isGPUSuitable = false;
+		}
+
+		// Check if the surface is adequate
+		{
+			std::vector<video::ISurface::SFormat> surfaceFormats =
+				gpu->getAvailableFormatsForSurface(surface.get());
+
+			std::vector<video::ISurface::E_PRESENT_MODE> presentModes =
+				gpu->getAvailablePresentModesForSurface(surface.get());
+
+			// Todo(achal): Probably should make a ISurface::SCapabilities
+			// struct for this as a wrapper for VkSurfaceCapabilitiesKHR
+			// nbl::video::ISurface::SCapabilities surfaceCapabilities = ;
+			VkSurfaceCapabilitiesKHR surfaceCapabilities;
+			vkGetPhysicalDeviceSurfaceCapabilitiesKHR(gpu->getInternalObject(),
+				vk_surface->m_surface, &surfaceCapabilities);
+
+			if ((surfaceCapabilities.maxImageCount != 0) && (SC_IMG_COUNT > surfaceCapabilities.maxImageCount)
+				|| (surfaceFormats.empty()) || (presentModes.empty()))
+			{
+				isGPUSuitable = false;
+			}
+
+			// Todo(achal): Probably a more sophisticated way to choose these
+			surfaceFormat = surfaceFormats[0];
+			presentMode = presentModes[0];
+			preTransform = static_cast<nbl::video::ISurface::E_SURFACE_TRANSFORM_FLAGS>(surfaceCapabilities.currentTransform);
+		}
+
+		if (isGPUSuitable) // find the first suitable GPU
+			break;
+	}
+	assert((graphicsFamilyIndex != ~0u) && (presentFamilyIndex != ~0u));
+
+	video::ILogicalDevice::SCreationParams deviceCreationParams;
+	if (graphicsFamilyIndex == presentFamilyIndex)
+	{
+		deviceCreationParams.queueParamsCount = 1u;
+		imageSharingMode = asset::ESM_EXCLUSIVE;
+	}
+	else
+	{
+		deviceCreationParams.queueParamsCount = 2u;
+		imageSharingMode = asset::ESM_CONCURRENT;
+	}
+
+	queueFamilyIndices = core::make_refctd_dynamic_array<QueueFamilyIndicesArrayType>(deviceCreationParams.queueParamsCount);
+	{
+		const uint32_t temp[] = { graphicsFamilyIndex, presentFamilyIndex };
+		for (uint32_t i = 0u; i < deviceCreationParams.queueParamsCount; ++i)
+			(*queueFamilyIndices)[i] = temp[i];
+	}
+
+	std::vector<video::ILogicalDevice::SQueueCreationParams> queueCreationParams(deviceCreationParams.queueParamsCount);
+	for (uint32_t i = 0u; i < deviceCreationParams.queueParamsCount; ++i)
+	{
+		queueCreationParams[i].familyIndex = (*queueFamilyIndices)[i];
+		queueCreationParams[i].count = 1u;
+		queueCreationParams[i].flags = static_cast<video::IGPUQueue::E_CREATE_FLAGS>(0);
+		const float priority = 1.f;
+		queueCreationParams[i].priorities = &priority;

```

Comment:
that float will go out of scope (UB abound) by the time you get into `deviceCreationParams`, it needs to be declared outside of the `for` loop

Reply:
yes ofcourse